sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -415,10 +465,19 @@ def release(self, instruction_id):
     Resets the ``BundleProcessor`` and moves it from the active to the
     inactive cache.
     """
-    descriptor_id, processor = self.active_bundle_processors.pop(instruction_id)
+    with self._lock:
+      self.known_not_running_instruction_ids[instruction_id] = True
+      while len(self.known_not_running_instruction_ids
+                ) > MAX_KNOWN_NOT_RUNNING_INSTRUCTIONS:
+        self.known_not_running_instruction_ids.popitem()
+      descriptor_id, processor = (
+          self.active_bundle_processors.pop(instruction_id))
+
+    # Make sure that we reset the processor while not holding the lock.
     processor.reset()
-    self.last_access_times[descriptor_id] = time.time()
-    self.cached_bundle_processors[descriptor_id].append(processor)
+    with self._lock:
+      self.last_access_times[descriptor_id] = time.time()
+      self.cached_bundle_processors[descriptor_id].append(processor)
 
   def shutdown(self):
     """
```

Comment:
Not sure if any locking needs to be done in this method. Perhaps `known_not_running_instruction_ids` should be cleared?

It seems that this is called when no more requests will be processed (end of SdkHarness.run()), so this should be fine.

Reply:
From my understanding I also think it should be fine since it is on shutdown.